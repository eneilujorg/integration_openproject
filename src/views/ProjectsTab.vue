<!--
  - @copyright Copyright (c) 2022 Kiran Parajuli <kiran@jankaritech.com>
  -
  - @author Kiran Parajuli <kiran@jankaritech.com>
  -
  - @license GNU AGPL version 3 or any later version
  -
  - This program is free software: you can redistribute it and/or modify
  - it under the terms of the GNU Affero General Public License as
  - published by the Free Software Foundation, either version 3 of the
  - License, or (at your option) any later version.
  -
  - This program is distributed in the hope that it will be useful,
  - but WITHOUT ANY WARRANTY; without even the implied warranty of
  - MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  - GNU Affero General Public License for more details.
  -
  - You should have received a copy of the GNU Affero General Public License
  - along with this program. If not, see <http://www.gnu.org/licenses/>.
  -
  -->

<template>
	<div class="projects">
		<SearchInput
			:file-info="fileInfo"
			@saved="onSaved" />
		<div v-if="isLoading" class="icon-loading" />
		<div v-else-if="workpackages.length > 0" id="openproject-linked-workpackages">
			<div class="existing-relations">
				{{ t('integration_openproject', 'Existing relations:') }}
			</div>
			<WorkPackage
				v-for="workpackage in workpackages"
				:key="workpackage.id"
				:workpackage="workpackage" />
		</div>
		<EmptyContent v-else
			id="openproject-empty-content"
			:state="state"
			:request-url="requestUrl"
			:admin-config-status="adminConfigStatus" />
	</div>
</template>

<script>
import EmptyContent from '../components/tab/EmptyContent'
import WorkPackage from '../components/tab/WorkPackage'
import axios from '@nextcloud/axios'
import { generateUrl } from '@nextcloud/router'
import SearchInput from '../components/tab/SearchInput'
import { loadState } from '@nextcloud/initial-state'
import { workpackageHelper } from '../utils/workpackageHelper'

export default {
	name: 'ProjectsTab',
	components: {
		EmptyContent,
		SearchInput,
		WorkPackage,
	},
	data: () => ({
		error: '',
		fileInfo: { },
		state: 'loading',
		workpackages: [],
		requestUrl: loadState('integration_openproject', 'request-url'),
		adminConfigStatus: loadState('integration_openproject', 'admin-config-status'),
	}),
	computed: {
		isLoading() {
			return this.state === 'loading'
		},
	},
	methods: {
		/**
		 * updates current resource
		 *
		 * @param {object} fileInfo file information
		 */
		async update(fileInfo) {
			this.fileInfo = fileInfo
			this.workpackages = []
			this.state = 'loading'
			await this.fetchWorkpackages(this.fileInfo.id)
		},
		/**
		 * Reset the current view to its default state
		 */
		resetState() {
			this.error = ''
			this.state = 'loading'
		},
		onSaved(data) {
			this.workpackages.push(data)
		},
		async fetchWorkpackages(fileId) {
			const req = {}
			const url = generateUrl('/apps/integration_openproject/work-packages?fileId=' + fileId)
			try {
				const response = await axios.get(url, req)
				if (!Array.isArray(response.data)) {
					this.state = 'failed-fetching-workpackages'
				} else {
					// empty data means there are no workpackages linked
					if (response.data.length > 0) {
						for (let workPackage of response.data) {
							workPackage = await workpackageHelper.getAdditionalMetaData(workPackage)
							this.workpackages.push(workPackage)
						}
					}
					this.state = 'ok'
				}
			} catch (error) {
				if (error.response && error.response.status === 401) {
					this.state = 'no-token'
				} else if (error.response && error.response.status === 404) {
					this.state = 'connection-error'
				} else if (error.response && error.response.status === 500) {
					this.state = 'error'
				} else {
					this.state = 'failed-fetching-workpackages'
				}
			}
		},
	},
}
</script>

<style scoped lang="scss">
.projects {
	height: 100% !important;
	width: 100%;

	.existing-relations {
		text-align: left;
		font-weight: bold;
		font-size: 0.8rem;
		padding: 12px;
	}

	.center-content {
		display: flex;
		align-items: center;
		justify-content: center;
	}

	.title {
		font-size: 2rem;
		font-weight: 600;
		padding-bottom: 0;
	}

	.subtitle {
		padding-top: 0;
		font-size: 1.2rem;
	}

	.icon-loading:after {
		top: 140%;
	}
}
</style>
