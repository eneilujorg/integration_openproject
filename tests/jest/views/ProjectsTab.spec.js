/* jshint esversion: 8 */

import { shallowMount, mount, createLocalVue } from '@vue/test-utils'
import ProjectsTab from '../../../src/views/ProjectsTab'
import axios from '@nextcloud/axios'
import * as initialState from '@nextcloud/initial-state'
import workPackagesSearchResponse from '../fixtures/workPackagesSearchResponse.json'

jest.mock('@nextcloud/dialogs')
jest.mock('@nextcloud/l10n', () => ({
	translate: jest.fn((app, msg) => msg),
	getLanguage: jest.fn(() => ''),
}))

global.OC = {
	dialogs: {
		confirmDestructive: jest.fn(),
		YES_NO_BUTTONS: 70,
	},
}

jest.mock('@nextcloud/axios')
const localVue = createLocalVue()

describe('ProjectsTab.vue Test', () => {
	let wrapper
	const loadingIndicatorSelector = '.icon-loading'
	const emptyContentSelector = '#openproject-empty-content'
	const workPackagesSelector = '#openproject-linked-workpackages'
	const existingRelationSelector = '.existing-relations'
	const searchInputStubSelector = 'searchinput-stub'
	const linkedWorkpackageSelector = '.workpackage'
	const workPackageUnlinkSelector = '.linked-workpackages--workpackage--unlink'

	beforeEach(() => {
		// eslint-disable-next-line no-import-assign
		initialState.loadState = jest.fn(() => 'https://openproject/oauth/')
		wrapper = shallowMount(ProjectsTab, { localVue })
	})
	describe('search input existence', () => {
		it('should not exist if the request url is not valid', async () => {
			await wrapper.setData({
				requestUrl: false,
			})
			expect(wrapper.find(searchInputStubSelector).exists()).toBeFalsy()
		})
		it('should exist if the request url is valid', async () => {
			await wrapper.setData({
				requestUrl: 'https://open.project/',
			})
			expect(wrapper.find(searchInputStubSelector).exists()).toBeTruthy()
		})
	})
	describe('loading icon', () => {
		it('shows the loading icon during "loading" state', async () => {
			wrapper.setData({ state: 'loading' })
			await localVue.nextTick()
			expect(wrapper.find(loadingIndicatorSelector).exists()).toBeTruthy()
		})
		it('does not show the empty content message during "loading" state', async () => {
			wrapper.setData({ state: 'loading' })
			await localVue.nextTick()
			expect(wrapper.find(emptyContentSelector).exists()).toBeFalsy()
		})
		it.each(['ok', 'error'])('makes the loading icon disappear on state change', async (state) => {
			wrapper.setData({ state: 'loading' })
			await localVue.nextTick()
			expect(wrapper.find(loadingIndicatorSelector).exists()).toBeTruthy()
			wrapper.setData({ state })
			await localVue.nextTick()
			expect(wrapper.find(loadingIndicatorSelector).exists()).toBeFalsy()
		})
		it('sets the state to "loading" on update', async () => {
			const err = new Error()
			err.response = { status: 404 }
			axios.get.mockRejectedValueOnce(err)
			wrapper.setData({ state: 'ok' })
			expect(wrapper.vm.state).toBe('ok')
			wrapper.vm.update({ id: 123 })
			expect(wrapper.vm.state).toBe('loading')
		})
	})
	describe('empty message', () => {
		it.each(['no-token', 'ok', 'error'])('shows the empty message when state is other than loading', async (state) => {
			wrapper.setData({ state })
			await localVue.nextTick()
			expect(wrapper.find(emptyContentSelector).exists()).toBeTruthy()
		})
	})
	describe('fetchWorkpackages', () => {
		it.each([
			{ HTTPStatus: 400, AppState: 'failed-fetching-workpackages' },
			{ HTTPStatus: 401, AppState: 'no-token' },
			{ HTTPStatus: 402, AppState: 'failed-fetching-workpackages' },
			{ HTTPStatus: 404, AppState: 'connection-error' },
			{ HTTPStatus: 500, AppState: 'error' },
		])('sets states according to HTTP error codes', async (cases) => {
			const err = new Error()
			err.response = { status: cases.HTTPStatus }
			axios.get.mockRejectedValueOnce(err)
			await wrapper.vm.update({ id: 123 })
			expect(wrapper.vm.state).toBe(cases.AppState)
		})
		it.each([
			null,
			'string',
			undefined,
			[{ // missing id
				subject: 'subject',
				_links: {
					status: {
						href: '/api/v3/statuses/12',
						title: 'open',
					},
					type: {
						href: '/api/v3/types/6',
						title: 'Task',
					},
					assignee: {
						href: '/api/v3/users/1',
						title: 'Bal Bahadur Pun',
					},
					project: { title: 'a big project' },
				},
			}],
			[{ // empty subject
				id: 123,
				subject: '',
				_links: {
					status: {
						href: '/api/v3/statuses/12',
						title: 'open',
					},
					type: {
						href: '/api/v3/types/6',
						title: 'Task',
					},
					assignee: {
						href: '/api/v3/users/1',
						title: 'Bal Bahadur Pun',
					},
					project: { title: 'a big project' },
				},
			}],
			[{ // missing subject
				id: 123,
				_links: {
					status: {
						href: '/api/v3/statuses/12',
						title: 'open',
					},
					type: {
						href: '/api/v3/types/6',
						title: 'Task',
					},
					assignee: {
						href: '/api/v3/users/1',
						title: 'Bal Bahadur Pun',
					},
					project: { title: 'a big project' },
				},
			}],
			[{ // missing _links.status.title
				id: 123,
				subject: 'my task',
				_links: {
					status: {
						href: '/api/v3/statuses/12',
					},
					type: {
						href: '/api/v3/types/6',
						title: 'Task',
					},
					assignee: {
						href: '/api/v3/users/1',
						title: 'Bal Bahadur Pun',
					},
					project: { title: 'a big project' },
				},
			}],
			[{ // missing project.title
				id: 123,
				subject: 'my task',
				_links: {
					status: {
						href: '/api/v3/statuses/12',
						title: 'open',
					},
					type: {
						href: '/api/v3/types/6',
						title: 'Task',
					},
					assignee: {
						href: '/api/v3/users/1',
						title: 'Bal Bahadur Pun',
					},
					project: { },
				},
			}],
		])('sets the "failed-fetching-workpackages" state on invalid responses', async (testCase) => {
			axios.get
				.mockImplementationOnce(() => Promise.resolve({
					status: 200,
					data: testCase,
				}))
				// mock for color requests, it should not fail because of the missing mock
				.mockImplementation(() => Promise.resolve({ status: 200, data: [] }))
			await wrapper.vm.update({ id: 123 })
			expect(wrapper.vm.state).toBe('failed-fetching-workpackages')
		})
		it('sets the "ok" state on empty response', async () => {
			axios.get
				.mockImplementation(() => Promise.resolve({ status: 200, data: [] }))
			await wrapper.vm.update({ id: 123 })
			expect(wrapper.vm.state).toBe('ok')
		})
		it.each([
			{ statusColor: { color: '#A5D8FF' }, typeColor: { color: '#00B0F0' } },
			{ statusColor: { }, typeColor: { } },
		])('shows the linked workpackages', async (testCase) => {
			wrapper = mountWrapper()
			const axiosGetSpy = jest.spyOn(axios, 'get')
				.mockImplementationOnce(() => Promise.resolve({
					status: 200,
					data: [{
						id: 123,
						subject: 'my task',
						_links: {
							status: {
								href: '/api/v3/statuses/12',
								title: 'open',
							},
							type: {
								href: '/api/v3/types/6',
								title: 'Task',
							},
							assignee: {
								href: '/api/v3/users/1',
								title: 'Bal Bahadur Pun',
							},
							project: { title: 'a big project' },
						},
					},
					{
						id: 589,
						subject: 'नेपालमा IT उद्योग बनाउने',
						_links: {
							status: {
								href: '/api/v3/statuses/2',
								title: 'प्रगति हुदैछ',
							},
							type: {
								href: '/api/v3/types/16',
								title: 'Epic',
							},
							assignee: {
								href: '/api/v3/users/13',
								title: 'कुमारी नेपाली',
							},
							project: { title: 'नेपालको विकास गर्ने' },
						},
					}],
				}))
				// mock for color requests
				.mockImplementationOnce(() => Promise.resolve(
					{ status: 200, data: testCase.statusColor })
				)
				.mockImplementationOnce(() => Promise.resolve(
					{ status: 200, data: testCase.typeColor })
				)
			await wrapper.vm.update({ id: 789 })
			expect(axiosGetSpy).toBeCalledWith(
				'http://localhost/apps/integration_openproject/work-packages?fileId=789',
				{}
			)
			expect(axiosGetSpy).toBeCalledWith(
				'http://localhost/apps/integration_openproject/statuses/12',
			)
			expect(axiosGetSpy).toBeCalledWith(
				'http://localhost/apps/integration_openproject/types/6',
			)
			expect(wrapper.vm.state).toBe('ok')
			const workPackages = wrapper.find(workPackagesSelector)
			expect(wrapper.find(existingRelationSelector).exists()).toBeTruthy()
			expect(workPackages.exists()).toBeTruthy()
			expect(workPackages).toMatchSnapshot()
		})
	})
	describe('onSave', () => {
		it('shows the just linked workpackage', async () => {
			wrapper = mountWrapper()
			await wrapper.vm.onSaved(workPackagesSearchResponse[0])
			const workPackages = wrapper.find(workPackagesSelector)
			expect(wrapper.find(existingRelationSelector).exists()).toBeTruthy()
			expect(workPackages.exists()).toBeTruthy()
			expect(workPackages).toMatchSnapshot()
		})
	})
	describe('when the work package is clicked', () => {
		it('opens work package in open project', async () => {
			axios.get
				.mockImplementationOnce(() => Promise.resolve({
					status: 200,
					data: 'http://openproject',
				}))
			window.open = jest.fn()
			wrapper = mountWrapper()
			await wrapper.setData({
				workpackages: workPackagesSearchResponse,
			})
			await localVue.nextTick()
			await wrapper.find(linkedWorkpackageSelector).trigger('click')
			await localVue.nextTick()
			expect(window.open).toHaveBeenCalledTimes(1)
			expect(window.open).toHaveBeenCalledWith(
				'http://openproject/projects/15/work_packages/1'
			)
		})
	})
	describe('when workpackage unlink button is clicked', () => {
		it('should display a confirmation dialog box', async () => {
			wrapper = mountWrapper()
			await wrapper.setData({
				workpackages: workPackagesSearchResponse,
			})
			await localVue.nextTick()
			await expect(wrapper.find(workPackageUnlinkSelector).exists()).toBeTruthy()
			await wrapper.find(workPackageUnlinkSelector).trigger('click')
			await localVue.nextTick()
			expect(OC.dialogs.confirmDestructive).toHaveBeenCalledTimes(1)
			expect(OC.dialogs.confirmDestructive).toHaveBeenCalledWith(
				'Are you sure you want to unlink the work package?',
				'Confirm unlink',
				{ cancel: 'Cancel', confirm: 'Unlink', confirmClasses: 'error', type: 70 },
				expect.any(Function),
				true
			)
		})
	})

	describe('unlinkWorkPackage', () => {
		it('should unlink the work package', async () => {
			const axiosGetSpy = jest.spyOn(axios, 'get')
				.mockImplementationOnce(() => Promise.resolve({
					status: 200,
					data: [{
						_type: 'FileLink',
						id: 66,
						createdAt: '2022-04-06T05:14:24Z',
						updatedAt: '2022-04-06T05:14:24Z',
						originData: {
							id: '6',
							name: 'welcome.txt',
							mimeType: 'text/plain',
							createdAt: '1970-01-01T00:00:00Z',
							lastModifiedAt: '2022-03-30T07:39:56Z',
							createdByName: '',
							lastModifiedByName: '',
						},
						_links: {
							delete: {
								href: '/api/v3/file_links/66',
								method: 'delete',
							},
						},
					}],
				}))
			const axiosDeleteSpy = jest.spyOn(axios, 'delete').mockImplementationOnce(() => Promise.resolve(
				{ status: 200 })
			)
			wrapper = mountWrapper()
			await wrapper.vm.unlinkWorkPackage(15, 6)
			expect(axiosGetSpy).toBeCalledWith(
				'http://localhost/apps/integration_openproject/work-packages/15/file-links'
			)
			expect(axiosDeleteSpy).toBeCalledWith('http://localhost/apps/integration_openproject/file-links/66')
			axiosGetSpy.mockRestore()
			axiosDeleteSpy.mockRestore()
		})

		it.each([
			{ HTTPStatus: 401, state: 'no-token' },
			{ HTTPStatus: 404, state: 'error' },
			{ HTTPStatus: 500, state: 'error' },
		])('sets states according to HTTP error codes', async (cases) => {
			const err = new Error()
			err.response = { status: cases.HTTPStatus }
			axios.get.mockRejectedValueOnce(err)
			wrapper = mountWrapper()
			await wrapper.vm.unlinkWorkPackage(15, 6)
			expect(wrapper.vm.state).toBe(cases.state)
		})
	})
})

function mountWrapper() {
	return mount(ProjectsTab, {
		localVue,
		mocks: {
			t: (msg) => msg,
			generateUrl() {
				return '/'
			},
		},
		stubs: {
			SearchInput: true,
			Avatar: true,
		},
		data: () => ({
			error: '',
			state: 'ok',
			fileInfo: {},
			workpackages: [],
			requestUrl: 'something',
		}),
	})
}
