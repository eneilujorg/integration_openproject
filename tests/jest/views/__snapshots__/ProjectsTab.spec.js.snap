// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectsTab.vue Test fetchWorkpackages adds every work-package only once 1`] = `
<div class="" id="openproject-linked-workpackages">
  <div class="existing-relations">
    integration_openproject
  </div>
  <div class="linked-workpackages">
    <div class="linked-workpackages--workpackage">
      <div class="workpackage linked-workpackages--workpackage--item" id="workpackage-123">
        <div class="row">
          <div class="row__status">
            <div class="row__status__title">
              open
            </div>
          </div>
          <div class="row__workpackage">
            #123 - a big project
          </div>
        </div>
        <div class="row">
          <div class="row__subject"><span class="row__subject__type">
				Task
			</span>
            my task
          </div>
        </div>
        <div class="row">
          <div class="row__assignee">
            <div class="row__assignee__avatar">
              <avatar-stub size="23" url="http://localhost/apps/integration_openproject/avatar?userId=1&amp;userName=Bal Bahadur Pun" user="Bal Bahadur Pun" display-name="Bal Bahadur Pun" class="item-avatar"></avatar-stub>
            </div>
            <div class="row__assignee__assignee">
              Bal Bahadur Pun
            </div>
          </div>
        </div>
      </div> <button data-v-2a98ba3b="" rel="nofollow noreferrer noopener" class="action-item action-item--single icon-noConnection linked-workpackages--workpackage--unlinkactionbutton undefined has-tooltip" aria-label="" icon="icon-noConnection" data-original-title="">
        <!---->



        <span data-v-2a98ba3b="" aria-hidden="true" hidden="hidden"><li data-v-ef50c406="" class="action linked-workpackages--workpackage--unlinkactionbutton" data-v-2a98ba3b=""><button data-v-ef50c406="" aria-label="" type="button" class="action-button focusable"><span data-v-ef50c406="" class="action-button__icon icon-noConnection"></span> <span data-v-ef50c406="" class="action-button__text">Unlink WorkPackage</span>
        <!---->
      </button></li></span></button>
    </div>
    <div class=""></div>
  </div>
</div>
`;

exports[`ProjectsTab.vue Test fetchWorkpackages sets the "error" state if the admin config is not okay 1`] = `
<div class="projects projects--empty">
  <!---->
  <div class="empty-content" id="openproject-empty-content">
    <div class="empty-content--wrapper">
      <div class="empty-content--icon"><img src="" alt="error"></div>
      <!---->
      <div class="empty-content--connect-button">
        <div class="oauth-connect--message">
          Some OpenProject integration application settings are not working. Please contact your Nextcloud administrator.
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ProjectsTab.vue Test fetchWorkpackages shows the linked workpackages 1`] = `
<div class="" id="openproject-linked-workpackages">
  <div class="existing-relations">
    integration_openproject
  </div>
  <div class="linked-workpackages">
    <div class="linked-workpackages--workpackage">
      <div class="workpackage linked-workpackages--workpackage--item" id="workpackage-123">
        <div class="row">
          <div class="row__status" style="background-color: rgb(165, 216, 255);">
            <div class="row__status__title">
              open
            </div>
          </div>
          <div class="row__workpackage">
            #123 - a big project
          </div>
        </div>
        <div class="row">
          <div class="row__subject"><span class="row__subject__type" style="color: rgb(0, 176, 240);">
				Task
			</span>
            my task
          </div>
        </div>
        <div class="row">
          <div class="row__assignee">
            <div class="row__assignee__avatar">
              <avatar-stub size="23" url="http://localhost/apps/integration_openproject/avatar?userId=1&amp;userName=Bal Bahadur Pun" user="Bal Bahadur Pun" display-name="Bal Bahadur Pun" class="item-avatar"></avatar-stub>
            </div>
            <div class="row__assignee__assignee">
              Bal Bahadur Pun
            </div>
          </div>
        </div>
      </div> <button data-v-2a98ba3b="" rel="nofollow noreferrer noopener" class="action-item action-item--single icon-noConnection linked-workpackages--workpackage--unlinkactionbutton undefined has-tooltip" aria-label="" icon="icon-noConnection" data-original-title="">
        <!---->



        <span data-v-2a98ba3b="" aria-hidden="true" hidden="hidden"><li data-v-ef50c406="" class="action linked-workpackages--workpackage--unlinkactionbutton" data-v-2a98ba3b=""><button data-v-ef50c406="" aria-label="" type="button" class="action-button focusable"><span data-v-ef50c406="" class="action-button__icon icon-noConnection"></span> <span data-v-ef50c406="" class="action-button__text">Unlink WorkPackage</span>
        <!---->
      </button></li></span></button>
    </div>
    <div class="workpackage-seperator"></div>
  </div>
  <div class="linked-workpackages">
    <div class="linked-workpackages--workpackage">
      <div class="workpackage linked-workpackages--workpackage--item" id="workpackage-589">
        <div class="row">
          <div class="row__status" style="background-color: rgb(165, 216, 255);">
            <div class="row__status__title">
              प्रगति हुदैछ
            </div>
          </div>
          <div class="row__workpackage">
            #589 - नेपालको विकास गर्ने
          </div>
        </div>
        <div class="row">
          <div class="row__subject"><span class="row__subject__type" style="color: rgb(0, 176, 240);">
				Epic
			</span>
            नेपालमा IT उद्योग बनाउने
          </div>
        </div>
        <div class="row">
          <div class="row__assignee">
            <div class="row__assignee__avatar">
              <avatar-stub size="23" url="http://localhost/apps/integration_openproject/avatar?userId=13&amp;userName=कुमारी नेपाली" user="कुमारी नेपाली" display-name="कुमारी नेपाली" class="item-avatar"></avatar-stub>
            </div>
            <div class="row__assignee__assignee">
              कुमारी नेपाली
            </div>
          </div>
        </div>
      </div> <button data-v-2a98ba3b="" rel="nofollow noreferrer noopener" class="action-item action-item--single icon-noConnection linked-workpackages--workpackage--unlinkactionbutton undefined has-tooltip" aria-label="" icon="icon-noConnection" data-original-title="">
        <!---->



        <span data-v-2a98ba3b="" aria-hidden="true" hidden="hidden"><li data-v-ef50c406="" class="action linked-workpackages--workpackage--unlinkactionbutton" data-v-2a98ba3b=""><button data-v-ef50c406="" aria-label="" type="button" class="action-button focusable"><span data-v-ef50c406="" class="action-button__icon icon-noConnection"></span> <span data-v-ef50c406="" class="action-button__text">Unlink WorkPackage</span>
        <!---->
      </button></li></span></button>
    </div>
    <div class=""></div>
  </div>
</div>
`;

exports[`ProjectsTab.vue Test fetchWorkpackages shows the linked workpackages 2`] = `
<div class="" id="openproject-linked-workpackages">
  <div class="existing-relations">
    integration_openproject
  </div>
  <div class="linked-workpackages">
    <div class="linked-workpackages--workpackage">
      <div class="workpackage linked-workpackages--workpackage--item" id="workpackage-123">
        <div class="row">
          <div class="row__status">
            <div class="row__status__title">
              open
            </div>
          </div>
          <div class="row__workpackage">
            #123 - a big project
          </div>
        </div>
        <div class="row">
          <div class="row__subject"><span class="row__subject__type">
				Task
			</span>
            my task
          </div>
        </div>
        <div class="row">
          <div class="row__assignee">
            <div class="row__assignee__avatar">
              <avatar-stub size="23" url="http://localhost/apps/integration_openproject/avatar?userId=1&amp;userName=Bal Bahadur Pun" user="Bal Bahadur Pun" display-name="Bal Bahadur Pun" class="item-avatar"></avatar-stub>
            </div>
            <div class="row__assignee__assignee">
              Bal Bahadur Pun
            </div>
          </div>
        </div>
      </div> <button data-v-2a98ba3b="" rel="nofollow noreferrer noopener" class="action-item action-item--single icon-noConnection linked-workpackages--workpackage--unlinkactionbutton undefined has-tooltip" aria-label="" icon="icon-noConnection" data-original-title="">
        <!---->



        <span data-v-2a98ba3b="" aria-hidden="true" hidden="hidden"><li data-v-ef50c406="" class="action linked-workpackages--workpackage--unlinkactionbutton" data-v-2a98ba3b=""><button data-v-ef50c406="" aria-label="" type="button" class="action-button focusable"><span data-v-ef50c406="" class="action-button__icon icon-noConnection"></span> <span data-v-ef50c406="" class="action-button__text">Unlink WorkPackage</span>
        <!---->
      </button></li></span></button>
    </div>
    <div class="workpackage-seperator"></div>
  </div>
  <div class="linked-workpackages">
    <div class="linked-workpackages--workpackage">
      <div class="workpackage linked-workpackages--workpackage--item" id="workpackage-589">
        <div class="row">
          <div class="row__status">
            <div class="row__status__title">
              प्रगति हुदैछ
            </div>
          </div>
          <div class="row__workpackage">
            #589 - नेपालको विकास गर्ने
          </div>
        </div>
        <div class="row">
          <div class="row__subject"><span class="row__subject__type">
				Epic
			</span>
            नेपालमा IT उद्योग बनाउने
          </div>
        </div>
        <div class="row">
          <div class="row__assignee">
            <div class="row__assignee__avatar">
              <avatar-stub size="23" url="http://localhost/apps/integration_openproject/avatar?userId=13&amp;userName=कुमारी नेपाली" user="कुमारी नेपाली" display-name="कुमारी नेपाली" class="item-avatar"></avatar-stub>
            </div>
            <div class="row__assignee__assignee">
              कुमारी नेपाली
            </div>
          </div>
        </div>
      </div> <button data-v-2a98ba3b="" rel="nofollow noreferrer noopener" class="action-item action-item--single icon-noConnection linked-workpackages--workpackage--unlinkactionbutton undefined has-tooltip" aria-label="" icon="icon-noConnection" data-original-title="">
        <!---->



        <span data-v-2a98ba3b="" aria-hidden="true" hidden="hidden"><li data-v-ef50c406="" class="action linked-workpackages--workpackage--unlinkactionbutton" data-v-2a98ba3b=""><button data-v-ef50c406="" aria-label="" type="button" class="action-button focusable"><span data-v-ef50c406="" class="action-button__icon icon-noConnection"></span> <span data-v-ef50c406="" class="action-button__text">Unlink WorkPackage</span>
        <!---->
      </button></li></span></button>
    </div>
    <div class=""></div>
  </div>
</div>
`;

exports[`ProjectsTab.vue Test onSave shows the just linked workpackage 1`] = `
<div id="openproject-linked-workpackages">
  <div class="existing-relations">
    integration_openproject
  </div>
  <div class="linked-workpackages">
    <div class="linked-workpackages--workpackage">
      <div class="workpackage linked-workpackages--workpackage--item workpackage-transition" id="workpackage-1">
        <div class="row">
          <div class="row__status" style="background-color: blue;">
            <div class="row__status__title">
              in-progress
            </div>
          </div>
          <div class="row__workpackage">
            #1 - test
          </div>
        </div>
        <div class="row">
          <div class="row__subject"><span class="row__subject__type" style="color: red;">
				task
			</span>
            Organize work-packages
          </div>
        </div>
        <div class="row">
          <div class="row__assignee">
            <div class="row__assignee__avatar">
              <avatar-stub size="23" url="/server/index.php/apps/integration_openproject/avatar?userId=1&amp;userName=System" user="test" display-name="test" class="item-avatar"></avatar-stub>
            </div>
            <div class="row__assignee__assignee">
              test
            </div>
          </div>
        </div>
      </div> <button data-v-2a98ba3b="" rel="nofollow noreferrer noopener" class="action-item action-item--single icon-noConnection linked-workpackages--workpackage--unlinkactionbutton undefined has-tooltip" aria-label="" icon="icon-noConnection" data-original-title="">
        <!---->



        <span data-v-2a98ba3b="" aria-hidden="true" hidden="hidden"><li data-v-ef50c406="" class="action linked-workpackages--workpackage--unlinkactionbutton" data-v-2a98ba3b=""><button data-v-ef50c406="" aria-label="" type="button" class="action-button focusable"><span data-v-ef50c406="" class="action-button__icon icon-noConnection"></span> <span data-v-ef50c406="" class="action-button__text">Unlink WorkPackage</span>
        <!---->
      </button></li></span></button>
    </div>
    <div class=""></div>
  </div>
</div>
`;
